core/cas-server-core-services/src/main/java/org/apereo/cas/services/DomainServicesManager.java

```diff
@@ -0,0 +1,210 @@
+     */
+    private final Pattern domainPattern = RegexUtils.createPattern("^(https?|imaps?)://([^:/]+)/i");
+
+    /**
+     * Instantiates a new default services manager impl.
+     *
+     * @param serviceRegistryDao the service registry dao
+     */
+    public DomainServicesManager(final ServiceRegistryDao serviceRegistryDao) {
+        this.serviceRegistryDao = serviceRegistryDao;
+    }
+
+    @Audit(action = "DELETE_SERVICE", actionResolverName = "DELETE_SERVICE_ACTION_RESOLVER",
+            resourceResolverName = "DELETE_SERVICE_RESOURCE_RESOLVER")
+    @Override
+    public synchronized RegisteredService delete(final long id) {
+        final RegisteredService service = findServiceBy(id);
+        if (service != null) {
+            this.serviceRegistryDao.delete(service);
+            this.services.remove(id);
+            this.domains.get(getDomain(service.getServiceId())).remove(service);
+            publishEvent(new CasRegisteredServiceDeletedEvent(this, service));
+        }
+        return service;
+    }
+
+    @Override
+    public RegisteredService findServiceBy(final Service service) {
+        return service != null ? findServiceBy(service.getId()) : null;
+    }
+
+    @Override
+    public Collection<RegisteredService> findServiceBy(final Predicate<RegisteredService> predicate) {
+        return services.values().stream()
+                .filter(predicate)
+                .sorted()
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public <T extends RegisteredService> T findServiceBy(final Service serviceId, final Class<T> clazz) {
+        return findServiceBy(serviceId.getId(), clazz);
+    }
+
+    @Override
+    public <T extends RegisteredService> T findServiceBy(final String serviceId, final Class<T> clazz) {
+        return getServicesForDomain(getDomain(serviceId)).stream()
+                .filter(s -> s.getClass().isAssignableFrom(clazz) && s.matches(serviceId))
+                .map(clazz::cast)
+                .findFirst()
+                .orElse(null);
+    }
+
+    @Override
+    public RegisteredService findServiceBy(final long id) {
+        final RegisteredService r = this.services.get(id);
+
+        try {
+            return r == null ? null : r.clone();
+        } catch (final CloneNotSupportedException e) {
+            return r;
+        }
+    }
+
+    @Override
+    public RegisteredService findServiceBy(final String serviceId) {
+        String domain = serviceId != null ? getDomain(serviceId) : "";
+        domain = domains.containsKey(domain) ? domain : "default";
+        return domains.get(domain)
+                .stream()
+                .filter(s -> s.matches(serviceId))
+                .findFirst().orElse(null);
+    }
+
+    @Override
+    public Collection<RegisteredService> getAllServices() {
+        return Collections.unmodifiableCollection(services.values().stream().sorted().collect(Collectors.toList()));
```

Comment:
Please avoid ` Collections.unmodifiableCollection`. If you want to ensure the original collection remains in place, simply use `CollectionUtils.wrap`. Unmodifiable collections mess up serialization and all kinds of weirdness ensues.